### PR TITLE
Document parallel reporters

### PR DESCRIPTION
**Pull request by gaborcsardi**

### CODE REVIEW

File: vignettes/parallel.Rmd [hunk 1]
```diff
@@ -58,3 +58,96 @@ There is some overhead associated with running tests in parallel:
 This overhead generally means that if you have many test files that take a short amount of time, you're unlikely to see a huge benefit by using parallel tests.
 For example, testthat itself takes about 10s to run tests in serial, and 8s to run the tests in parallel.
 That's because most tests take less than a second to run, and the slowest test (which takes 3s) is started last.
+
+## Reporters
+
+### Default reporters
+
+See `default_reporter()` for how testthat selects the default reporter.
+In short, testthat selects `ProgressReporter` for non-parallel and `ParallelProgressReporter` for parallel tests by default.
```
Comment (kevinushey) [inline vignettes/parallel.Rmd:67]: Is this true for `R CMD check` as well? (Is the CheckReporter no longer used?)

File: vignettes/parallel.Rmd [hunk 2]
```diff
@@ -58,3 +58,96 @@ There is some overhead associated with running tests in parallel:
...
+
+-   Reporters report all information about a file at once, unless they support *parallel updates*.
+    E.g.
+    `ProgressReporter` does not update its display until a test file is complete.
+
+-   The standard output and standard error, i.e. `print()`, `cat()`, `message()`, etc. output from the test files are lost currently.
+    If you want to use `cat()` or `message()` for print-debugging test cases, then the best is to temporarily run tests sequentially, by changing the `Config` entry in `DESCRIPTION` or selecting a non-parallel reporter, e.g. the `CheckReporter`:
+
+    ``` {.r}
+    devtools::test(filter = "badtest", repoter = "check")
```
Comment (kevinushey) [inline vignettes/parallel.Rmd:97]:
```suggestion
    devtools::test(filter = "badtest", reporter = "check")
```

File: vignettes/parallel.Rmd [hunk 3]
```diff
@@ -58,3 +58,96 @@ There is some overhead associated with running tests in parallel:
...
+## Reporters
+
+### Default reporters
+
+See `default_reporter()` for how testthat selects the default reporter.
+In short, testthat selects `ProgressReporter` for non-parallel and `ParallelProgressReporter` for parallel tests by default.
+
+### Parallel support
+
+Most reporters support parallel tests.
```
Comment (kevinushey) [inline vignettes/parallel.Rmd:71]: For parallel tests, is it just with each test file run in separate sub-processes (in parallel)? Or could different individual tests (`test_that()` blocks) be run in parallel too?

File: vignettes/parallel.Rmd (hunk 1 above)
Comment (gaborcsardi) [reply vignettes/parallel.Rmd:67]: No, you are right, `test_check()` uses `CheckReporter` by default. Let me rephrase this.

File: vignettes/parallel.Rmd (hunk 3 above)
Comment (gaborcsardi) [reply vignettes/parallel.Rmd:71]: No, it is by file.

File: vignettes/parallel.Rmd [hunk 4]
```diff
@@ -58,3 +58,96 @@ There is some overhead associated with running tests in parallel:
...
+  self$capabilities$parallel_updates <- TRUE
+  ...
+}
+...
+```
+
+For these reporters, testthat does not cache the messages from the subprocesses.
+Instead, when a message comes in:
+
+-   It calls the `$start_file()` method, letting the reporter know which file the following calls apply to.
```
Comment (kevinushey) [inline vignettes/parallel.Rmd:147]: Is there any worry about interleaving of `$start_file()` or the following `$start_test()` / `$add_result()` from multiple sub-processes? How is synchronization handled there? (seems like in theory the parent process could receive multiple `$start_file()` messages at once?)

Comment (gaborcsardi) [reply vignettes/parallel.Rmd:147]: No, there is no such worry, because the subsequent `start_file()` messages are not real, they are not coming from the subprocess. The subprocess only sends one `start_file()` at the beginning. 

The logic in the main process prepends a `start_file()` before each subsequent message from the subprocess, to let the reporter know that the next message(s) are for a certain file.

File: vignettes/parallel.Rmd (hunk 3 above)
Comment (gaborcsardi) [reply vignettes/parallel.Rmd:71]: Clarified this at the beginning.

**Comment by gaborcsardi** [comment]
@hadley I think this is ready for review.

**Review by hadley — APPROVED** [review]
